---
title: "Individual-Specific Network for Stress-Related Psychiatric Disorders"
sub-title: Explore GDS formats
output:
  html_document:
    df_print: paged
editor_options: 
  chunk_output_type: console
---

```{r}
library(dplyr)
library(data.table)

library(gdsfmt)
library(SNPRelate)

library(arrow)

library(ggplot2)
library(RColorBrewer)
brewer_pallete <- brewer.pal(n = 8, name = "Dark2")
```

```{r}
setwd("/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/")

source("~/kul/dex-stim-human-array-isns/code/00_functions/load_gds_files.R")
```


```{r}
knitr::opts_knit$set(root.dir = "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/input/")
```

# 1. DNAm

## Baseline 

```{r load-dnam}
treatment <- "veh"
gds_fn   <- paste0("/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/input/dnam/gds/methyl_beta_mtrx_corrected_for_cov",  
                   "_", treatment, ".gds") 

dnam_gds_file <- openfn.gds(gds_fn)

dnam_mtrx <- read.gdsn(index.gdsn(dnam_gds_file, "beta_mtrx"))
colnames(dnam_mtrx)   <- read.gdsn(index.gdsn(dnam_gds_file, "cpg_id"))
rownames(dnam_mtrx)   <- read.gdsn(index.gdsn(dnam_gds_file, "sample_id"))

closefn.gds(dnam_gds_file)
```

## Check dimensionality of DNAm mtrx

```{r}
dim(dnam_mtrx)
```

## Check summary of the first 5 CpGs

```{r}
summary(dnam_mtrx[, 1:5])
```

## 5 CpGs for the first 10 samples

```{r}
dnam_mtrx[1:10, 1:5]
```
# 2. Genotype

## 2.1. Explore transformed object

```{r load-snp-gds}
snps_gds <- snpgdsOpen("/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/input/snps/gds/dex_geno_imputed.gds")

# snpset <- snpgdsLDpruning(snps_gds, ld.threshold=0.2, maf = 0.05, slide.max.bp = 1000000L, slide.max.n = 100, method = "corr")

geno_obj  <- snpgdsGetGeno(snps_gds, with.id = T)
snp_chr   <- read.gdsn(index.gdsn(snps_gds, "snp.chromosome"))

snpgdsClose(snps_gds)

snp_mtrx   <- geno_obj$genotype
colnames(snp_mtrx)   <- geno_obj$snp.id
rownames(snp_mtrx)   <- geno_obj$sample.id
```

## 2.2. Check dimensionality of SNP mtrx

```{r}
dim(snp_mtrx)
```

## 2.3. Check summary of the first 5 SNPs

```{r}
summary(snp_mtrx[, 1:5])
```

## 2.4. SNP for the first 10 samples

```{r}
snp_mtrx[1:10, 1:5]
```
## 2.5. Compare SNP mtrx with the matrix that was used for eQTL/meQTL analysis: transformed using plink

### Load SNP matrix that was used for eQTL/meQTL analysis

```{r load-snp-qtl}
snp_mpip <- read_delim_arrow("/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/data/integrative/matrixEQTL/snp_mtrx.csv", delim = ";")
```

### Align columns and rows

```{r}
snp_mtrx <- snp_mtrx[rownames(snp_mtrx) %in% colnames(snp_mpip)[-1],]
dim(snp_mpip)
```

### Compare GDS SNP mtrx with QTL SNP mtrx

```{r}
print("Sample IDs for missing values from GDS mtrx: ")
snp_mtrx[is.na(snp_mtrx[, 2]), 2]
```

```{r}
rs <- "rs3131971"

print(paste0("QTL matrix: table for ", rs))
table(t(snp_mpip[snp_mpip$SNP == rs, -1]))

print(paste0("GDS matrix: table for ", rs))
table(snp_mtrx[, rs])
```

```{r}
rs <- colnames(snp_mtrx)[7894]

print(paste0("QTL matrix: table for ", rs))
table(t(snp_mpip[snp_mpip$SNP == rs, -1]))

print(paste0("GDS matrix: table for ", rs))
table(snp_mtrx[, rs])
```



```{r}
t_snp_mtrx <- t(snp_mtrx)
dim(t_snp_mtrx)
t_snp_mtrx[["SNP"]] <- rownames(t_snp_mtrx)

fwrite(t_snp_mtrx, 
       "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/data/integrative/matrixEQTL/snp_mtrx_with_na.csv",
       quote = F, row.names = T, sep = ";")
```

```{r}
m <- fread("/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/data/integrative/matrixEQTL/snp_mtrx_with_na.csv")

m <- m[, -1]
```

# 3. Explore GDS SNP file obtained from Oxford gen format

```{r load-snp-gds}
setwd("/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/input")
snps_gds_fn <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/input/snps/gds/dex_geno_imputed_maf_ld_pruned_from_gen.gds"

snp_mtrx <- LoadGenotype(snps_gds_fn)
```

```{r}
summary(snp_mtrx[, 1:5])
```
```{r}
snp_mtrx <- snp_mtrx[rownames(snp_mtrx) %in% colnames(snp_mpip)[-1],]
```

```{r}
rs <- "rs3131971"

print(paste0("QTL matrix: table for ", rs))
table(t(snp_mpip[snp_mpip$SNP == rs, -1]))

print(paste0("GDS matrix: table for ", rs))
table(snp_mtrx[, rs])
```

```{r}
rs <- colnames(snp_mtrx)[4]

print(paste0("QTL matrix: table for ", rs))
table(t(snp_mpip[snp_mpip$SNP == rs, -1]))

print(paste0("GDS matrix: table for ", rs))
table(snp_mtrx[, rs])
```

# 4. Variants distribution across chromosomes

## 4.1. SNPs

```{r load-snp-coordintaes, include = F}
setwd("/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/")

snps_gds_fn <- paste0("input/snps/ld_pruned/gds/dex_geno_imputed_maf_ld_pruned_from_gen.gds")
snp_loc     <- LoadGenotypeCoordinates(snps_gds_fn) %>% data.frame() %>% setDT()
```

```{r}
snp_loc$chr <- as.numeric(snp_loc$chr)
```

```{r}
net_loc_count <- snp_loc %>%
  group_by(chr) %>%
  tally() %>%
  mutate(freq = n / sum(n)) %>% 
  mutate(chr = factor(chr, levels = 1:22)) %>% setDT()
```

```{r}
plt.size <- 14
ggplot(net_loc_count, aes(x = chr, y = freq, fill = chr)) + 
  geom_bar(stat = "identity", position = "stack") +
  geom_text(data = net_loc_count[freq > 0.02], 
            aes(label = scales::percent(freq, accuracy = 0.1), y = freq), 
            stat = "identity", position = position_stack(vjust = 0.95), size =  plt.size / 3,
            color = "black") +
    geom_text(data = net_loc_count, 
            aes(label = n, y = freq), 
            stat = "identity", position = position_stack(vjust = 1.05), size = plt.size / 3,
            color = "black") +
  scale_y_continuous(labels = scales::percent) +
  labs(title = "",
       x = "Chromosomes",
       y = "SNP frequency") +
  theme(legend.title = element_blank(), 
        legend.position = "none",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = plt.size),
        axis.title = element_text(size = plt.size, colour = "black"),
        axis.text.x = element_text(angle = 0, hjust = 0.5, colour = "black", size = plt.size),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  scale_fill_manual(values = alpha(rep(brewer_pallete, 3), 0.5)) +
  scale_color_manual(values = alpha(rep(brewer_pallete, 3), 0.75))
```

## 4.2. CpGs

```{r load-cpg-coordintaes, include = F}
setwd("/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/")

treatment   <- "veh"
dnam_gds_fn <- paste0("input/dnam/mad_filtered/gds/methyl_beta_mtrx_corrected_for_cov_mad80_filtered_", treatment, ".gds")
cpg_loc <- LoadMethylCoordinates(dnam_gds_fn) %>% data.frame() %>% setDT()
```

```{r}
cpg_loc$chr <- as.numeric(cpg_loc$chr)
```

```{r}
net_loc_count <- cpg_loc %>%
  group_by(chr) %>%
  tally() %>%
  mutate(freq = n / sum(n)) %>% 
  mutate(chr = factor(chr, levels = 1:22)) %>% setDT()
```

```{r}
plt.size <- 14
ggplot(net_loc_count, aes(x = chr, y = freq, fill = chr)) + 
  geom_bar(stat = "identity", position = "stack") +
  geom_text(data = net_loc_count[freq > 0.02], 
            aes(label = scales::percent(freq, accuracy = 0.1), y = freq), 
            stat = "identity", position = position_stack(vjust = 0.95), size =  plt.size / 3,
            color = "black") +
    geom_text(data = net_loc_count, 
            aes(label = n, y = freq), 
            stat = "identity", position = position_stack(vjust = 1.05), size = plt.size / 3,
            color = "black") +
  scale_y_continuous(labels = scales::percent) +
  labs(title = "",
       x = "Chromosomes",
       y = "CpG frequency") +
  theme(legend.title = element_blank(), 
        legend.position = "none",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = plt.size),
        axis.title = element_text(size = plt.size, colour = "black"),
        axis.text.x = element_text(angle = 0, hjust = 0.5, colour = "black", size = plt.size),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  scale_fill_manual(values = alpha(rep(brewer_pallete, 3), 0.5)) +
  scale_color_manual(values = alpha(rep(brewer_pallete, 3), 0.75))
```




