---
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r include = F}
library(dplyr)
library(data.table)
library(gdsfmt)

source("~/kul/dex-stim-human-array-isns/code/00_functions/load_gds_files.R")
source("~/kul/dex-stim-human-array-isns/code/00_functions/get_chord_diagramm.R")

cbPalette <- c("grey", "#0072B2")
brewer_pallete <- brewer.pal(n = 8, name = "Dark2")
```

```{r include = F}
rslt_dir <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/results/5_fold_cv/all/"
setwd(rslt_dir)
```

```{r load-data, include = F}
treatment <- "dex"
dex_global_net <- readRDS(paste0(rslt_dir, treatment, "/networks/smccnet_global_chr_all.rds"))

treatment <- "veh"
veh_global_net <- readRDS(paste0(rslt_dir, treatment, "/networks/smccnet_global_chr_all.rds"))
```

# Test for significance

```{r}
# treatment <- "dex"
# dex_global_net <- readRDS(paste0(rslt_dir, treatment, "/networks/smccnet_individual_MPIPSYKL_014030_chr_all.rds"))# smccnet_global_chr_all.rds"))
# 
# treatment <- "veh"
# veh_global_net <- readRDS(paste0(rslt_dir, treatment, "/networks/smccnet_individual_MPIPSYKL_014030_chr_all.rds")) # smccnet_global_chr_all.rds"))
# 
# dex_global_net <- dex_global_net$network
# veh_global_net <- veh_global_net$network
```


# Overlap between dex and veh features

```{r}
dex_features <- rownames(dex_global_net) # 5'388
veh_features <- rownames(veh_global_net) # 4'519

olap <- intersect(dex_features, veh_features) # 758
olap_cpgs <- olap[grepl("cg", olap, ignore.case = T)] # 566
olap_snps <- olap[grepl("rs", olap, ignore.case = T)] # 192
```


## CpG level

```{r out.width = "95%", include = T, eval = T}
euler.tbl <- euler(list(dex = dex_features[grepl("cg", dex_features, ignore.case = T)], 
                        veh = veh_features[grepl("cg", veh_features, ignore.case = T)]))

plot(euler.tbl,
     quantities = list(type = c("counts", "percent")),
     fills = alpha(cbPalette[c(2, 1)], 0.5),
     col = alpha(cbPalette[c(2, 1)], 1),
     labels =  c("GR-stimulated", "Baseline"))
```

## SNP level

```{r out.width = "95%", include = T, eval = T}
euler.tbl <- euler(list(dex = dex_features[grepl("rs", dex_features, ignore.case = T)], 
                        veh = veh_features[grepl("rs", veh_features, ignore.case = T)]))

plot(euler.tbl,
     quantities = list(type = c("counts", "percent")),
     fills = alpha(cbPalette[c(2, 1)], 0.5),
     col = alpha(cbPalette[c(2, 1)], 1),
     labels =  c("GR-stimulated", "Baseline"))
```

```{r load-cpg-coordintaes, include = F}
setwd("/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/")

treatment   <- "veh"
dnam_gds_fn <- paste0("input/dnam/mad_filtered/gds/methyl_beta_mtrx_corrected_for_cov_mad80_filtered_", treatment, ".gds")
cpg_loc <- LoadMethylCoordinates(dnam_gds_fn) %>% data.frame() %>% setDT()
```

```{r load-snp-coordintaes, include = F}
setwd("/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/")

snps_gds_fn <- paste0("input/snps/ld_pruned/gds/dex_geno_imputed_maf_ld_pruned_from_gen.gds")
snp_loc     <- LoadGenotypeCoordinates(snps_gds_fn) %>% data.frame() %>% setDT()
```

```{r create-tbl-for-plt, include = F}
dex_net_cpg_loc <- cpg_loc[CpG_ID %in% colnames(dex_global_net)] %>% 
  mutate(ID = CpG_ID, Type = "CpG", treatment = "GR-stimulated") %>% dplyr::select(-c(CpG_ID))
dex_net_snp_loc <- snp_loc[SNP %in% colnames(dex_global_net)] %>% 
  mutate(ID = SNP, Type = "SNP", treatment = "GR-stimulated") %>% dplyr::select(-SNP)

veh_net_cpg_loc <- cpg_loc[CpG_ID %in% colnames(veh_global_net)] %>% 
  mutate(ID = CpG_ID, Type = "CpG", treatment = "Baseline") %>% dplyr::select(-c(CpG_ID))
veh_net_snp_loc <- snp_loc[SNP %in% colnames(veh_global_net)] %>% 
  mutate(ID = SNP, Type = "SNP", treatment = "Baseline") %>% dplyr::select(-SNP)

net_loc     <- rbind(dex_net_cpg_loc, dex_net_snp_loc, veh_net_cpg_loc, veh_net_snp_loc)
```

# Distribution of features, CpGs and SNPs across chromosmes and time points (not overlap)

```{r}
net_loc_count <- net_loc %>%
  group_by(chr, Type, treatment) %>%
  tally() %>%
  group_by(treatment) %>%
  mutate(freq = n / sum(n)) %>% 
  mutate(chr = factor(chr, levels = 1:22)) %>% setDT()
```

```{r}
ggplot(net_loc_count, aes(x = treatment, y = freq, fill = Type)) + 
  geom_bar(stat = "identity", position = "stack") +
  geom_text(data = net_loc_count[freq > 0.01], 
            aes(label = scales::percent(freq, accuracy = 1), y = freq), 
            stat = "identity", position = position_stack(vjust = 0.5), size = 3,
            color = "black") +
  scale_y_continuous(labels = scales::percent) +
  facet_wrap(.~chr, nrow = 1, strip.position = "top") +
  labs(title = "",
       x = "",
       y = "Variant frequency") +
  theme(legend.title = element_blank(), 
        legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = plt.size),
        axis.title = element_text(size = plt.size, colour = "black"),
        axis.text.x = element_text(angle = 90, hjust = 0.5, colour = "black"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  scale_fill_manual(values = alpha(brewer_pallete, 0.25)) +
  scale_color_manual(values = alpha(brewer_pallete, 0.75))
```

<!-- ## SNPs -->

```{r include = F}
ggplot(net_loc_count[Type == "SNP"], aes(x = chr, y = freq, fill = treatment, color = treatment)) + 
  geom_bar(stat = "identity", position = "stack") +
  geom_text(aes(label = scales::percent(freq, accuracy = 0.1), y = freq), 
            stat = "identity", position = position_stack(vjust = 0.5), size = 3,
            color = "black") +
  scale_y_continuous(labels = scales::percent) +
#  facet_wrap(.~Type, nrow = 2) +
  labs(title = "",
       x = "Chromosomes",
       y = "SNP frequency") +
  theme(legend.title = element_blank(), 
        legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  scale_color_manual(values = alpha(cbPalette, 0.75)) +
  scale_fill_manual(values = alpha(cbPalette, 0.75))
```

# Nr of CpGs per SNP 
```{r veh-snp-cpg_cnt, include = F}
# Take only QTLs
thrsh <- 0.7

veh_qtl_net <- veh_global_net[rownames(veh_global_net) %in% snp_loc$SNP, colnames(veh_global_net) %in% cpg_loc$CpG_ID,]  %>%
  as.matrix() %>% as.data.frame()
  
  # Transform net to long format 
veh_data_long <- veh_qtl_net %>% 
  rownames_to_column %>%
  gather(key = 'key', value = 'value', -rowname) %>%
  mutate(CpG_ID = key, SNP = rowname) %>%
  dplyr::select(SNP, CpG_ID, value)

veh_qtl_cnt <- veh_data_long[veh_data_long$value >= thrsh, ] %>% 
  group_by(SNP) %>%
  tally() %>% 
  mutate(treatment = "Baseline")
```

```{r dex-snp-cpg_cnt, include = F}
# Take only QTLs
dex_qtl_net <- dex_global_net[rownames(dex_global_net) %in% snp_loc$SNP, colnames(dex_global_net) %in% cpg_loc$CpG_ID,]  %>%
  as.matrix() %>% as.data.frame()
  
  # Transform net to long format 
dex_data_long <- dex_qtl_net %>% 
  rownames_to_column %>%
  gather(key = 'key', value = 'value', -rowname) %>%
  mutate(CpG_ID = key, SNP = rowname) %>%
  dplyr::select(SNP, CpG_ID, value)

dex_qtl_cnt <- dex_data_long[dex_data_long$value >= thrsh, ] %>% 
  group_by(SNP) %>%
  tally() %>% 
  mutate(treatment = "GR-stimulated")
```

```{r}
qtl_cnt <- rbind(veh_qtl_cnt, dex_qtl_cnt)

plt.size <- 12
ggplot(qtl_cnt, aes(x = treatment, y = n, color = treatment, fill = treatment)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  theme(legend.position = "bottom", # c(0.9, 0.9), 
        legend.title = element_blank(), 
        legend.text = element_text(size = plt.size),
        panel.grid.major = element_blank(), 
        panel.background = element_blank(),
        plot.title = element_text(size = plt.size), 
        axis.title = element_text(size = plt.size),
        axis.text = element_text(size = plt.size, colour = "black")) + 
  labs(title = "",
       x = "", y = "Nr. of CpGs per SNP") +
  scale_color_manual(values = cbPalette) +
  scale_fill_manual(values = alpha(cbPalette, 0.25))
```

# Chord plot of meQTLs

## Basline

```{r}
snp_max <- veh_qtl_cnt[order(veh_qtl_cnt$n), ][10:15,]
CustomChordDiagram(veh_global_net, cpg_loc, snp_loc[SNP %in% snp_max$SNP,], thrsh = 0)
```


```{r}
treatment <- "Baseline"

# snp_max <- veh_qtl_cnt[order(veh_qtl_cnt$n, decreasing = T), ][10, ]
# cdm <- veh_global_net_flatten_sign[column %in% snp_max, ]
cdm <- veh_data_long[order(veh_data_long$value, decreasing = T),]
cdm <- left_join(cdm, net_loc[treatment == treatment, c("ID", "chr", "pos")], by = c("CpG_ID" = "ID")) %>% 
  mutate(cpg_chr = paste0("chr", chr), cpg_pos = pos) %>% dplyr::select(-c(chr, pos))
cdm <- left_join(cdm, net_loc[treatment == treatment, c("ID", "chr", "pos")], by = c("SNP" = "ID")) %>% 
  mutate(snp_chr = paste0("chr", chr), snp_pos = pos) %>% dplyr::select(-c(chr, pos)) %>% unique()
head(cdm) #  rs28513678 cg16400794 1.0000000    chr1 27709850    chr3   3776878

veh_cdm <- cdm

cdm$cpg_pos <- as.numeric(cdm$cpg_pos)
cdm$snp_pos <- as.numeric(cdm$snp_pos)

cdm <- setDT(cdm)

bed1 = cdm[1:500, .(cpg_chr, cpg_pos, value, CpG_ID)] %>% select(chr = cpg_chr, start = cpg_pos, end = cpg_pos, value, cpg = CpG_ID)
bed2 = cdm[1:500, .(snp_chr, snp_pos, value, SNP)] %>% select(chr = snp_chr, start = snp_pos, end = snp_pos, value, snp = SNP)

color.gradient <- function(x, colors=c("yellow", "green", "purple"), colsteps=140) {
  return( colorRampPalette(colors) (colsteps) [ findInterval(x, seq(min(x),max(x), length.out=colsteps)) ] )
}

bed1$col <-color.gradient(bed1$value)
# bed_corr <- list(bed1[cor > 0], bed1[cor < 0])
lgd_links <- Legend(at = signif(seq(min(bed1$value), max(bed1$value), by = 0.1), 0.01), 
                            col_fun = color.gradient, 
    title_position = "topleft", title = "value")

# cytoband <- read.cytoband(species = "hg19")$df
# cytoband <- cytoband[!(cytoband$V1 %in% c("chrX, chrY")), ]
circos.clear()
# circos.initialize()
# circos.par("gap.degree" = rep(c(1, 20), 1))
# circos.initializeWithIdeogram(cytoband, plotType = NULL)
# circos.genomicIdeogram(cytoband)
# circos.genomicLabels(bed1,  labels.column = 6, col = bed1$col, side = "outside")
circos.initializeWithIdeogram(chromosome.index = paste0("chr", 1:22))
# circos.genomicRainfall(bed_corr, mode = "min", pch = 16, cex = 0.8, col = c("#FF000080", "#0000FF80"))
circos.genomicDensity(bed1, col = brewer_pallete[1])
circos.genomicDensity(bed2, col =  brewer_pallete[2])
circos.genomicLink(bed1, bed2, col = bed1$col, border = NA)
draw(lgd_links, x = unit(10, "mm"), y = unit(10, "mm"), just = c("left", "bottom"))
```

## GR-stimulated

```{r}
snp_max <- dex_qtl_cnt[order(dex_qtl_cnt$n), ][1:10,]
CustomChordDiagram(dex_global_net, cpg_loc, snp_loc[SNP %in% snp_max$SNP,], thrsh = 0)
```


```{r}
treatment <- "GR-stimulated"

# snp_max <- veh_qtl_cnt[order(veh_qtl_cnt$n, decreasing = T), ][10, ]
# cdm <- veh_global_net_flatten_sign[column %in% snp_max, ]
cdm <- dex_data_long[order(dex_data_long$value, decreasing = T),]
cdm <- left_join(cdm, net_loc[treatment == treatment, c("ID", "chr", "pos")], by = c("CpG_ID" = "ID")) %>% 
  mutate(cpg_chr = paste0("chr", chr), cpg_pos = pos) %>% dplyr::select(-c(chr, pos))
cdm <- left_join(cdm, net_loc[treatment == treatment, c("ID", "chr", "pos")], by = c("SNP" = "ID")) %>% 
  mutate(snp_chr = paste0("chr", chr), snp_pos = pos) %>% dplyr::select(-c(chr, pos)) %>% unique()
head(cdm) # rs61904579 cg15978899 0.9385598    chr6  30882994   chr11 56472268

dex_cdm <- cdm

cdm$cpg_pos <- as.numeric(cdm$cpg_pos)
cdm$snp_pos <- as.numeric(cdm$snp_pos)

cdm <- setDT(cdm)

bed1 = cdm[1:500, .(cpg_chr, cpg_pos, value, CpG_ID)] %>% select(chr = cpg_chr, start = cpg_pos, end = cpg_pos, value, cpg = CpG_ID)
bed2 = cdm[1:500, .(snp_chr, snp_pos, value, SNP)] %>% select(chr = snp_chr, start = snp_pos, end = snp_pos, value, snp = SNP)

color.gradient <- function(x, colors=c("yellow", "green", "purple"), colsteps=140) {
  return( colorRampPalette(colors) (colsteps) [ findInterval(x, seq(min(x),max(x), length.out=colsteps)) ] )
}

bed1$col <-color.gradient(bed1$value)
# bed_corr <- list(bed1[cor > 0], bed1[cor < 0])
lgd_links <- Legend(at = signif(seq(min(bed1$value), max(bed1$value), by = 0.1), 0.01), 
                            col_fun = color.gradient, 
    title_position = "topleft", title = "value")

# cytoband <- read.cytoband(species = "hg19")$df
# cytoband <- cytoband[!(cytoband$V1 %in% c("chrX, chrY")), ]
circos.clear()
# circos.initialize()
# circos.par("gap.degree" = rep(c(1, 20), 1))
# circos.initializeWithIdeogram(cytoband, plotType = NULL)
# circos.genomicIdeogram(cytoband)
# circos.genomicLabels(bed1,  labels.column = 6, col = bed1$col, side = "outside")
circos.initializeWithIdeogram(chromosome.index = paste0("chr", 1:22))
# circos.genomicRainfall(bed_corr, mode = "min", pch = 16, cex = 0.8, col = c("#FF000080", "#0000FF80"))
circos.genomicDensity(bed1, col = brewer_pallete[1])
circos.genomicDensity(bed2, col =  brewer_pallete[2])
circos.genomicLink(bed1, bed2, col = bed1$col, border = NA)
draw(lgd_links, x = unit(10, "mm"), y = unit(10, "mm"), just = c("left", "bottom"))
```

## Examples

```{r, include = F, eval = T}
source("~/mpip/projects/dex-stim-human-array/code/integrative/util.R")

snp.df              <- LoadGenotype() %>% setDT()
methyl.beta.dex.df  <- LoadMethylBeta(treatment = "dex") %>% setDT()
methyl.beta.veh.df  <- LoadMethylBeta(treatment = "veh") %>% setDT()
```

### Both

```{r out.width = "95%", include = T, eval = T}
dex_cdm <- paste(dex_cdm$SNP, dex_cdm$CpG_ID, sep = "-")
ceh_cdm <- paste(veh_cdm$SNP, veh_cdm$CpG_ID, sep = "-")

intersect_cdm   <- inner_join(dex_cdm, veh_cdm, by = c("SNP", "CpG_ID", "meQTL"))
(selected_meqtl <- intersect_cdm[25,]) # rs4881713 cg01572694 0.8213771     chr17  46657555     chr11  50327851

ProcessGetBoxPlot(methyl.beta.veh.df, methyl.beta.dex.df, snp.df, selected_meqtl, fdr.thr = 0.05)
```

### Baseline

```{r out.width = "95%", include = T, eval = T}
veh_cdm_unique  <- veh_cdm[!(veh_cdm$meQTL %in% intersect_cdm$meQTL),]
(selected_meqtl <- veh_cdm_unique[6, ]) # rs10838661 cg16400794 0.9564631    chr1 27709850   chr11 5830617

ProcessGetBoxPlot(methyl.beta.veh.df, methyl.beta.dex.df, snp.df, selected_meqtl, fdr.thr = 0.05)
```

### GR-stimulated

```{r out.width = "95%", include = T, eval = T}
dex_cdm_unique  <- dex_cdm[!(dex_cdm$meQTL %in% intersect_cdm$meQTL),]
(selected_meqtl <- dex_cdm_unique[450, ]) #rs12224066 cg21592059 0.7915405    chr8 74332256   chr11 50228358 rs12224066-cg21592059

ProcessGetBoxPlot(methyl.beta.veh.df, methyl.beta.dex.df, snp.df, selected_meqtl, fdr.thr = 0.05)
```

####------------
# 4. Check how many cis / trans

# modules_features <- features[do.call(c, modules)]
# 
# cpg_loc[CpG_ID %in% modules_features]
# snp_loc[SNP %in% modules_features]

####------------

# 5. Coord plot 

## 5.1. Baseline

net <- veh_global_net %>% as.matrix() %>% as.data.frame()
net <- veh_global_net[!rownames(veh_global_net) %in% olap,
                      !colnames(veh_global_net) %in% olap] %>% as.matrix() %>% as.data.frame()
CustomChordDiagram(net, cpg_loc, snp_loc, thrsh = 0.80)

## 5.2. Post-dexamethasone

net <- dex_global_net %>% as.matrix() %>% as.data.frame()
CustomChordDiagram(net, cpg_loc, snp_loc, thrsh = 0.25)

## 5.3. Net with olap features betweeb pre- and post-dex

### 5.3.1 Baseline

net <- veh_global_net[olap, olap] %>% as.matrix() %>% as.data.frame()
CustomChordDiagram(net, cpg_loc, snp_loc, thrsh = 0)





# Get CpGs' and SNPs' locations for only features present in the network "net"
net_cpg_loc <- cpg_loc[CpG_ID %in% colnames(net)] %>% mutate(ID = CpG_ID) %>% select(-CpG_ID)
net_snp_loc <- snp_loc[SNP %in% colnames(net)] %>% mutate(ID = SNP) %>% select(-SNP)
net_loc     <- rbind(net_cpg_loc, net_snp_loc)[match(rownames(net), ID)]
